### PR TITLE
Feature/cop 2135 inject variable in form json

### PR DESCRIPTION
**Pull request by kam-cop**
Functionality to interpolate following form json segments before processing them:

```
- pageOptions
- formComponents - excluding formComponents.data.url segment by copying over after interpolation

```

### CODE REVIEW

**Review by paulfgraham — CHANGES_REQUESTED** [review]

File: src/utils/FormPage/getFormPage.js [hunk 1]
```diff
@@ -38,4 +41,33 @@ const getFormPage = (pageOptions, formComponents, formData) => {
...
+  return JSON.parse(Utils.interpolateString(JSON.stringify(pageOptions), formData));
+}
+/**
+ * Interpolate 'Variable Expression' using formData excluding each component's data block - for local use only.
+ * @param {Array} formComponents The components defined at the top-level of the form.
+ * @param {object} formData The top-level form data, used for setting up components.
+ * @returns interpolated formComponents
+ */
+const interpolateFormComponents = (formComponents, formData) => {
+    return formComponents.map((formComponent) => {
```
[the remark this answers is not in the document]
Comment (kam-cop) [reply src/utils/FormPage/getFormPage.js]: Thanks fixed extra indentation

File: src/utils/FormPage/getFormPage.js [hunk 2]
```diff
@@ -38,4 +41,33 @@ const getFormPage = (pageOptions, formComponents, formData) => {
...
+      let formComponentDataUrl = undefined;
+      if (formComponent.data && formComponent.data.url) {
+        formComponentDataUrl = formComponent.data.url;
+      }
+      let  interpolatedFormComponent = JSON.parse(Utils.interpolateString(JSON.stringify(formComponent), formData));
```
[the remark this answers is not in the document]
Comment (kam-cop) [reply src/utils/FormPage/getFormPage.js]: Thanks fixed it

File: src/utils/FormPage/getFormPage.test.js [hunk 3]
```diff
@@ -11,7 +11,9 @@ describe('utils', () => {
         { id: 'a', fieldId: 'a', label: 'Alpha', type: 'text' },
         { id: 'b', fieldId: 'b', label: 'Bravo', type: 'text' },
         // eslint-disable-next-line no-template-curly-in-string
-        { id: 'c', fieldId: 'c', label: 'Charlie', type: 'radios', data: { url: '${urls.refData}/v3/charlies' } }
+        { id: 'c', fieldId: 'c', label: 'Charlie', type: 'radios', data: { url: '${urls.refData}/v3/charlies' } },
+        { id: 'd', fieldId: 'd', label: 'Roger ${currentUser.firstname}', type: 'text' },
```
[the remark this answers is not in the document]
Comment (kam-cop) [reply src/utils/FormPage/getFormPage.test.js:16]: Fixed

File: src/utils/FormPage/getFormPage.test.js [hunk 4]
```diff
@@ -11,7 +11,9 @@ describe('utils', () => {
...
+        { id: 'e', fieldId: 'e', label: 'Bravo ${currentUser.surname}', type: 'text' },
```
[the remark this answers is not in the document]
Comment (kam-cop) [reply src/utils/FormPage/getFormPage.test.js:18]: Fixed

File: src/utils/FormPage/getFormPage.test.js [hunk 5]
```diff
@@ -122,12 +124,48 @@ describe('utils', () => {
             { type: 'html', tagName: 'p', content: PAGE.components[0] },
             { ...PAGE.components[1], full_path: PAGE.components[1].fieldId },
             { type: 'html', tagName: 'p', content: PAGE.components[2] },
-            { use: 'c', ...C, cya_label: C.label, full_path: C.fieldId }
+            { use: 'c', ...C, cya_label: C.label, full_path: C.fieldId, data: { url: '${urls.refData}/v3/charlies' } }
```
[the remark this answers is not in the document]
Comment (kam-cop) [reply src/utils/FormPage/getFormPage.test.js:130]: Fixed

File: src/utils/FormPage/getFormPage.test.js [hunk 6]
```diff
@@ -122,12 +124,48 @@ describe('utils', () => {
...
           ],
           formData: {}
         });
       });
 
+      it('should interpolate and handle a page that references a form-level component with formData', () => {
+        const PAGE = {
+          title: 'Page ${postFixTitle}',
```
[the remark this answers is not in the document]
Comment (kam-cop) [reply src/utils/FormPage/getFormPage.test.js:139]: Fixed

File: src/utils/FormPage/getFormPage.test.js [hunk 7]
```diff
@@ -122,12 +124,48 @@ describe('utils', () => {
...
+            "Opening paragraph",
+            { type: 'heading', size: 'l', content: 'Page heading' },
+            "Closing paragraph",
+            { use: 'c' },
+            { use: 'd' },
+            "Kevin",
+            { use: 'e' },
+          ]
+        };
+        const a = FORM_DATA;
```
[the remark this answers is not in the document]
Comment (kam-cop) [reply src/utils/FormPage/getFormPage.test.js]: Fixed

**Review by paulfgraham — CHANGES_REQUESTED** [review]
Just realised you need to change the version number in `package.json` or this can't be released and published to npm...

File: package.json [hunk 8]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@ukhomeoffice/cop-react-form-renderer",
-  "version": "3.0.1",
+  "version": "3.0.2",
```
Comment (paulfgraham) [inline package.json:3]: I don't know what this is playing at... but the version already in `main` is `"3.0.2"` so can you move this up to `"3.0.3"`, please, @kam-cop. 🙂

**Review by paulfgraham — CHANGES_REQUESTED** [review]

**Pull request merged**